### DISCUSSION TITLE
Guide: How to Calculate your MFU

### DISCUSSION THREAD

**Discussion by muellerzr** [karpathy/nanochat#198] [General]
In the repo, the MFU calculation is tailored to **H100**'s. So for example, if you're running on a 5090, 4090, or 6000 {insert variation here} your MFU will look *really bad*.

We're going to walk through how to find the effective TFLOPs for your GPU and adjust the calculation.

TL;DR here's common values to replace in [base_train](https://github.com/karpathy/nanochat/blob/master/scripts/base_train.py#L288-L290) based on your hardware:

Replace 989e12 with:

| GPU Model                 | Non-Sparse TFLOPs |
|----------------------------|---------------|
| RTX 3090                  | 71e12         |
| RTX 4090                  | 165.2e12       |
| RTX 5090                  | 209.5e12       |
| A6000                     | 154.8e12       |
| RTX 6000 ADA              | 364e12        |
| RTX 6000 Blackwell Max-Q  | 438.9e12       |
| RTX 6000 Blackwell        | 503.8e12       |
| A100                      | 312e12        |
| H100                      | 989e12        |
| H200                      | 989e12        |

## How to find your Sparse TFLOPs

1. Search up your GPU model followed by "datasheet". [For example here's the Blackwell 6000 datasheet](https://images.nvidia.com/aem-dam/Solutions/geforce/blackwell/nvidia-rtx-blackwell-gpu-architecture.pdf)
2. Do a CTRL+F for "FP32 Accumulate". The number we want is "Peak BF16 Tensor TFLOPS with FP32 Accumulate", which correlates to PyTorch training. (BF16 if it supports it, FP16 if you need to go there e.g. Colab)
3. Grab the *smaller* of the two numbers presented (if there are two), this is your non-sparse TFLOPs
4. Adjust the code mentioned earlier, replacing the H100 TFLOP number for the appropriate one for your GPU!

<img width="786" height="1659" alt="Image" src="https://github.com/user-attachments/assets/2d11a7c8-db2c-4679-9c63-c576330d060e" />

**Comment by thepowerfuldeez** [comment]
I think you meant Non-sparse TFLOPs here Zach. Also suggest unified 209.5e12 numbers (hence tera =e12)

**Comment by muellerzr** [comment]
Thanks @thepowerfuldeez, adjusted :)

**Comment by karpathy** [comment]
Ty (possibly this is better under Discussion than Issue.). My hope is to integrate something along these lines:

https://github.com/pytorch/torchtitan/blob/main/torchtitan/tools/utils.py#L68

into nanochat. I'll probably get around to later today.

**Comment by muellerzr** [comment]
Yes 100%. Was thinking that too + everything consumer wise that supports BF16

(And no worries, not used to GH discussions but I’ll remember for next time!)

**Comment by jon-hotaisle** [comment]
@karpathy Add in this PR too if it is good: https://github.com/pytorch/torchtitan/pull/1909

**Comment by drQedwards** [comment]
# My recommendation thanks on Torch exception that Karpathy wants to integrate

So Good call, with a caveat that I learned on exception handling in #89 @karpathy  — this could follow the same early-return pattern I used in **[PR #89: Implement PMLL with memory loop](https://github.com/karpathy/nanochat/pull/89)**.

Instead of wrapping the whole function in a broad `try/except`, a shorter check-and-return block makes the flow cleaner and avoids masking unrelated exceptions. Something like:

```python
import shutil
import subprocess

def get_peak_flops(device_name: str) -> int | None:
    # Fail fast if lspci isn't available
    if not shutil.which("lspci"):
        return None

    result = subprocess.run(["lspci"], stdout=subprocess.PIPE, text=True)
    if not result.stdout:
        return None

    filtered_lines = [
        line for line in result.stdout.splitlines()
        if "NVIDIA" in line and "H100" in line
    ]
    if not filtered_lines:
        return None

    device_name = " ".join(filtered_lines)
    # TODO: map device_name -> numeric FLOPs (e.g., via a lookup table or nvml)
    return None
```

Same philosophy Nurik mentioned earlier — smaller `try` scopes and explicit early exits keep hardware detection and JSON parsing predictable, especially when integrating downstream logic like the memory loop in the PMLL branch.

*Happy to sync this pattern once #89 merges — it aligns nicely with nanochat’s overall design direction.*
